### DISCUSSION TITLE
Filament sensor freezes klipper[BUG]

### DISCUSSION THREAD

**Discussion by Trilis29** [bassamanator/Sovol-SV06-firmware#103] [General]
When filament runout sensor detects runout my printer just freezes and is unresponsive, is it correct hehaviour?

**Comment by bassamanator** [comment]
That's not the expected behavior. Check the FAQ section.

**Comment by Trilis29** [comment]
I have checkeded FAQ also checked wires connected correctly to PCB, when there is no filament, I can see near mainboard led light goes out at SV06, like something is shorting... or maybe I have bad schematics of this sensor and need to swap something

**Comment by bassamanator** [comment]
See the FAQ section for expected behaviour upon runout.

In any case, this is a wiring issue. Ensure that VCC, GRN, Signal wires lineup accordingly with hotend pcb.

**Reply by Trilis29** [reply]
This is the one I have, double checked the pinout and seems everything correct. I will try to switch VCC with SIGNAL
![image](https://github.com/bassamanator/Sovol-SV06-firmware/assets/51697921/3dbb1382-e394-43ed-b308-7d50a938905a)

**Reply by bassamanator** [reply]
How did you enable the filament sensor?

Forget about the sensor for a moment, can you complete a print?

**Reply by Trilis29** [reply]
enabled via misc-macros.cfg
`[gcode_macro _globals]
variable_filament_sensor_enabled: 1 # NOTE Enable(1) or disable(0) the filament sensor, if one is connected
variable_beeping_enabled: 1 # NOTE Enable(1) or disable(0) beeping everywhere except during gantry calibration
variable_bed_temp_over: 10 # NOTE Start print if bed temperature is over by this amount, otherwise wait for teperature drop
variable_kamp_enable: 1 # NOTE Enable(1) or disable(0) KAMP (adaptive mesh)
variable_pre_purge_prime_length: 1.40
gcode:
    # Don't delete this section`

**Reply by Trilis29** [reply]
it this correct way?

**Reply by bassamanator** [reply]
> Forget about the sensor for a moment, can you complete a print?

**Comment by 64bitcube** [comment]
For what it's worth and future records, while it may or may not have been the issue, I've been digging though my own runout sensor issues, and I can confirm that an incorrect wire arrangement on the sensor will indeed crash klipper.  On my Sv06, specifically, there was 3 outcomes amongst different possible wire combos:  1) Printer crashes and screen goes dark.  2) Printer freezes with screen on, but klipper still is crashed and needs to be restarted at the host device.  3)  Sensor  behaves 'normally', and doesn't crash klipper.  

Each of the crash config can occur on one state of the sensor or the other, depending on polarity.  So if @Trilis29's 'bug' is only triggering when the sensor is in the 'empty' state, I'm guessing (from my evidence) that the printer could only be booted (or, at least, the sensor could only successfully be plugged in and not crash) when the filament is already loaded.  If so, it's definitely a wire-coding issue.

**Reply by Trilis29** [reply]
I have taken off sensor completely, but I will try to switch polarity. By the way your 3rd outcome, can you show your wiring diagram? I suppose you are using same sensor.

**Reply by 64bitcube** [reply]
I'm using a kingroon sensor from Ali.  Attached is some quick terribly-framed photos of the wiring (and yes, the white pin is coming out in one photo.  I've addressed that ;)  ).  The order should be **Red-black-white**, depending on how you look at the connectors.  The connector to the sv06 should have a consistent orientation so there's only one shot of it, but note that the second shot of the sensor connector is from the 'side with the clip/click', and that metal of the pins is not visible (unlike the other side).

It's probably worth paying more attention to wiring position than wire color if you think we have the same sensors--I've heard that some of the knock-off sensors can occasionally have a wire color improperly coded right from the factory.  

  <details>
    <summary><u>Photos</u> (<i>click to expand</i>)</summary>
    
    
![20240129_213730_HDR](https://github.com/bassamanator/Sovol-SV06-firmware/assets/10658531/d45d4bc7-22ae-458d-96fa-ad83f350f4e8)
![20240129_213812_HDR](https://github.com/bassamanator/Sovol-SV06-firmware/assets/10658531/3abe53c0-83e8-4455-aad6-b08a60b9d3f9)

  </details>
</details>

Like I mentioned, I can't guarantee this setup will work perfectly either.  The current problem I need to fix is that klipper/mainsail is showing the sensor in it's opposite state, consistently.  Surprising given the description in the FAQ of how the filament sensor works, but maybe it's a trick of the 3 wire setups or something.  When I get some time in the next few days I'll sit down and mess with my .cfg and adjust what I think is the culprit.

Regardless, I hope it helps.

**Reply by 64bitcube** [reply]
To finish off my saga and hopefully help someone else later:  The filament sensors now work great on both my SV06s.  Turns out, despite my Kingroon sensors being on the same product listing on Ali, the signal varies between what color case you choose (either that or it's just a gamble).  So my regular black-shell sensor worked no problems, no edits, right away following the wire order I mentioned in the prior post.  The transparent-shelled case was what was giving me my trouble, because it's signal is reversed.  Needed to change the .cfg to remove  the '!' before the pin assignment for the sensor, and now it works like a charm too.  

  <details>
    <summary><u>Thumbnail of the sensor family</u> (<i>click to expand</i>)</summary>
    
    
![KINGROON-3D-Printer-Filament-Break-Detection-Module-With-1M-Cable-Run-out-Sensor-Material-Runout-Detector](https://github.com/bassamanator/Sovol-SV06-firmware/assets/10658531/87ba26a6-5057-4f08-8cc6-f232005a7245)

  </details>
</details>

Funny thing was that the behavior klipper/mainsail showed with the pre-fix transparent sensor was *only* announcing the 'filament load' event.  That had me scratching my head for a bit.  Another odd thing to note about the sensors:  It seems the cables that came with mine have their connector wire-colors *mirrored*.  As in, one 3-pin connector was in red-black-white, and the other was in white-black-red.  Is that typical?  Since I cut the wire down the middle to install my own jst-connector it didn't matter as much, but I could see this confusing someone if they don't ensure their from-the-factory connector is in the right/matching order first.
